text/0000-glimmer-components.md

```diff
@@ -0,0 +1,1309 @@
+practices as they stand. New users should see Glimmer components as the
+_default_, and should not ever have to write a classic component or see one in
+the main guides.
+
+Classic components will of course be widely used for some time however, so a
+classic section which includes conversion guides and relevant codemods should be
+made available in the guides, and maintained for as long as classic components
+are supported by Ember.
+
+Breaking down the public API of Glimmer components, we need to cover:
+
+* Native class syntax, including the `constructor` and class fields
+* Arguments
+* Lifecycle hooks and properties
+* Element modifiers
+
+### Native class syntax
+
+One of the benefits of native class syntax is that it is used outside of Ember,
+so as time goes on we will be able to assume there is more general knowledge of
+it, and provide links to documentation for it for users who are not familiar.
+During this transitionary period though we should add a more thorough primer of
+the syntax to our guides, and explicitly call out the differences between
+classic class syntax and native class syntax, including:
+
+* Usage of `constructor` in classes which _do not_ extend from classic classes.
+  Otherwise, always use `init`. This will be tricky, because even when using
+  native classes to extend from classic classes, you should still use `init`.
+
+* Nuances of class fields - they run _after_ `super`, and _before_ user code.
+
+* Benefits of class field initializers, and how they can be used to do much of
+  the work that would otherwise be done in the `constructor` or `init`
+
+* Expense of class fields - new objects and functions are created for every
+  instance, so users should also be careful with them.
+
+### Arguments
+
+For arguments, namespacing makes sense in general as an API choice and is common
+in other frameworks (`props` in React, etc.). We should be sure to cover this
+thoroughly for users who are used to classic components, but it shouldn't
+require too much explaining.
+
+Immutability will be a bigger sticking point in general, in particular the
+inability to provide default argument values. This is easy enough to work around
+using an `{{or}}` helper in templates:
+
+```hbs
+Hello, {{or @firstName "friend"}}!
+```
+
+Or a defaulting alias decorator in the class:
+
+```hbs
+Hello, {{this.firstName}}!
+```
+```js
+export default class Greeting extends GlimmerComponent {
+  @tracked
+  get firstName() {
+    return this.args.firstName || 'friend';
+  }
+}
+```
+
+But does add a bit of boilerplate to components. Users will also have to be
+careful when attempting to override these "defaults" in subclasses, since it is
+not as simple as overriding a class field or property. We should guide users to
+use template-only components to "curry" components when trying to provide
+defaults in subclasses instead, leveraging the outer HTML semantics of
+Glimmer components:
```

Comment:
There are alternatives, absolutely, and we should discuss them. Specifically, using defaulting aliases on the component class, and overriding the defaults in subclasses.

This solution is the "cleanest" one to me, and from what I understand will also be the most performant one at scale since TO components incur so little overhead. It does require us to add `...arguments` to make it truly ergonomic, and we're definitely punting on that until after Octane.